File: Csharp/Csharp/LinkedList.cs
using System;

namespace AlgorithmsDatastructures
{
    class SinglyLinkedListNode<T>
    {
        private T value;
        private SinglyLinkedListNode<T> next = null;

        public SinglyLinkedListNode(T value)
        {
            this.value = value;
        }

        public SinglyLinkedListNode(T value, SinglyLinkedListNode<T> next)
        {
            this.value = value;
            this.next = next;
        }

        public T Value
        {
            set { this.value = value; }
            get { return this.value; }
        }

        public SinglyLinkedListNode<T> Next
        {
            set { next = value; }
            get { return next; }
        }
    }

    class LinkedListNode<T>
    {
        private T value;
        private LinkedListNode<T> prev = null;
        private LinkedListNode<T> next = null;

        public LinkedListNode(T value)
        {
            this.value = value;
        }

        public LinkedListNode(T value, LinkedListNode<T> prev)
        {
            this.value = value;
            this.prev = prev;
        }

        public LinkedListNode(T value, LinkedListNode<T> prev, LinkedListNode<T> next)
        {
            this.value = value;
            this.prev = prev;
            this.next = next;
        }

        public T Value
        {
            set { this.value = value; }
            get { return value; }
        }

        public LinkedListNode<T> Prev
        {
            set { prev = value; }
            get { return prev; }
        }

        public LinkedListNode<T> Next
        {
            set { next = value; }
            get { return next; }
        }
    }

    class LinkedList<T>
    {
        private int length = 0;
        private LinkedListNode<T> start = null;
        private LinkedListNode<T> end = null;

        public int size()
        {
            return this.length;
        }

        public void append(T value)
        {
            LinkedListNode<T> n;

            if (start == null)
            {
                n = new LinkedListNode<T>(value);
                start = n;
            } else
            {
                n = new LinkedListNode<T>(value, end);
                end.Next = n;
            }

            end = n;
            length++;
        }

        public void insert(T value, int index)
        {
            if (index < 0 || index > length)
            {
                throw new IndexOutOfRangeException();
            }
            else if (index == length)
            {
                append(value);
            }
            else
            {
                LinkedListNode<T> n;

                if (index == 0)
                {
                    n = new LinkedListNode<T>(value, null, start);
                    start.Prev = n;
                    start = n;
                }
                else
                {
                    LinkedListNode<T> trgt = getNode(index);
                    n = new LinkedListNode<T>(value, trgt.Prev, trgt);
                    trgt.Prev.Next = n;
                    trgt.Prev = n;
                }

                length++;
            }
        }

        public LinkedListNode<T> getNode(int index)
        {
            if (index >= length)
            {
                throw new IndexOutOfRangeException();
            }
            else
            {
                LinkedListNode<T> cu = start;
                for (int i = 0; i < index; i++)
                {
                    cu = cu.Next;
                }

                return cu;
            }
        }

        public LinkedListNode<T> first()
        {
            return start;
        }

        public LinkedListNode<T> last()
        {
            return end;
        }

        public T get(int index)
        {
            try
            {
                return getNode(index).Value;
            } catch (IndexOutOfRangeException)
            {
                throw new IndexOutOfRangeException();
            }
        }

        public void remove(int index)
        {
            LinkedListNode<T> el = getNode(index);
            
            if (el != start)
            {
                el.Prev.Next = el.Next;
                el.Next.Prev = el.Prev;

                if (el == end)
                {
                    end = el.Prev;
                }
            } else
            {
                start = el.Next;
                el.Next.Prev = null;
            }

            length--;
        }

        public override string ToString()
        {
            string ret = String.Format("LinkedList[{0}]{{ ", typeof(T).Name);
            
            if (length > 16)
            {
                for (int i = 0; i < 8; i++)
                {
                    ret += get(i).ToString() + ", ";
                }

                ret += "..., ";

                for (int i = length-8; i < length; i++)
                {
                    ret += get(i).ToString();
                    ret += i == length - 1 ? " }" : ", ";
                }
            } else
            {
                for (int i = 0; i < length; i++)
                {
                    ret += get(i).ToString();
                    ret += i == length - 1 ? " }" : ", ";
                }
            }

            return ret;
        }
    }

    class SinglyLinkedList<T>
    {
        private int length = 0;
        private SinglyLinkedListNode<T> start = null;
        private SinglyLinkedListNode<T> end = null;

        public int size()
        {
            return this.length;
        }

        public void append(T value)
        {
            SinglyLinkedListNode<T> n = new SinglyLinkedListNode<T>(value);

            if (start == null)
            {
                start = n;
            }
            else
            {
                end.Next = n;
            }

            end = n;
            length++;
        }

        public void insert(T value, int index)
        {
            if (index < 0 || index > length)
            {
                throw new IndexOutOfRangeException();
            }
            else if (index == length)
            {
                append(value);
            }
            else
            {
                SinglyLinkedListNode<T> n;

                if (index == 0)
                {
                    n = new SinglyLinkedListNode<T>(value, start);
                    start = n;
                }
                else
                {
                    SinglyLinkedListNode<T> trgt = getNode(index-1);
                    n = new SinglyLinkedListNode<T>(value, trgt.Next);
                    trgt.Next = n;
                }

                length++;
            }
        }

        public SinglyLinkedListNode<T> getNode(int index)
        {
            if (index >= length)
            {
                throw new IndexOutOfRangeException();
            }
            else
            {
                SinglyLinkedListNode<T> cu = start;
                for (int i = 0; i < index; i++)
                {
                    cu = cu.Next;
                }

                return cu;
            }
        }

        public SinglyLinkedListNode<T> first()
        {
            return start;
        }

        public SinglyLinkedListNode<T> last()
        {
            return end;
        }

        public T get(int index)
        {
            try
            {
                return getNode(index).Value;
            }
            catch (IndexOutOfRangeException)
            {
                throw new IndexOutOfRangeException();
            }
        }

        public void remove(int index)
        {
            if (index == 0)
            {
                start = start.Next;
            } else
            {
                SinglyLinkedListNode<T> bel = getNode(index - 1);
                bel.Next = bel.Next.Next;

                if (index == length-1)
                {
                    end = bel;
                }
            }

            length--;
        }

        public override string ToString()
        {
            string ret = String.Format("LinkedList[{0}]{{ ", typeof(T).Name);

            if (length > 16)
            {
                for (int i = 0; i < 8; i++)
                {
                    ret += get(i).ToString() + ", ";
                }

                ret += "..., ";

                for (int i = length - 8; i < length; i++)
                {
                    ret += get(i).ToString();
                    ret += i == length - 1 ? " }" : ", ";
                }
            }
            else
            {
                for (int i = 0; i < length; i++)
                {
                    ret += get(i).ToString();
                    ret += i == length - 1 ? " }" : ", ";
                }
            }

            return ret;
        }
    }
}

File: Python/linked_list.py
class LinkedListNode(object):
    def __init__(self, value, prev=None, aftr=None):
        self.value = value
        self.prev = prev
        self.aftr = aftr

class LinkedList(object):
    def __init__(self):
        self.__start = None
        self.__end = None
        self.__length = 0

    def size(self):
        return self.__length

    def append(self, value):
        if self.__length == 0:
            n = LinkedListNode(value)
            self.__start = n
        else:
            n = LinkedListNode(value, self.__end)
            self.__end.aftr = n
        
        self.__end = n
        self.__length += 1
        
    def insert(self, value, index):
        if index < 0 or index > self.__length:
            raise IndexError()
        elif index == self.__length:
            self.append(value)
        else:
            if index == 0:
                n = LinkedListNode(value, None, self.__start)
                self.__start.prev = n
                self.__start = n
            else:
                trgt = self.get_node(index)
                n = LinkedListNode(value, trgt.prev, trgt)
                trgt.prev.aftr = n
                trgt.prev = n
                
        self.__length += 1

    def get_node(self, index):
        if index >= self.__length:
            raise IndexError()
        else:
            cu = self.__start
            for i in range(index):
                cu = cu.aftr
            return cu

    def first():
        return start

    def last():
        return end

    def get(self, index):
        try:
            return self.get_node(index).value
        except IndexError:
            raise IndexError()
        
    def remove(self, index):
        el = self.get_node(index)
        
        if el == self.__start:
            self.__start = self.__start.aftr
            self.__start.prev = None
        else:
            el.prev.aftr = el.aftr
            el.aftr.prev = el.prev
            
            if el == self.__end:
                self.__end = el.prev
        
        self.__length -= 1
        
    def __str__(self):
        ret = 'LinkedList{ '
        
        if self.__length > 16:
            for i in range(8):
                ret += str(self.get(i)) + ', '

            ret += '..., '
            
            for i in range(self.__length-8,self.__length):
                ret += str(self.get(i))
                ret += ' }' if i == self.__length - 1 else ', '
        else:
            for i in range(self.__length):
                ret += str(self.get(i))
                ret += ' }' if i == self.__length - 1 else ', '
                
        return ret

def main():
    lli = LinkedList()
    
    for i in range(32):
        lli.append(i)

    print(lli)

    lli.remove(0)
    lli.remove(1)

    print(lli)

    lli.insert(8, 1)
    
    print(lli)
    
if __name__ == '__main__':
    main()